### ISSUE TITLE
Bug - Type aliasing typing.Tuple is treated as if the empty Tuple is being aliased

### ISSUE THREAD

**Issue by ufl-taeber** [python/mypy#8715]
This is a bug report.

Type aliasing `typing.Tuple` is treated as if the empty tuple (`typing.Tuple[()]`) is being aliased.

Here is some sample code that converts Polar coordinates into Cartesian:
```python
import math
import typing

Tuple = typing.Tuple


def cartesian(r: int, θ: int) -> Tuple[float, float]:
    rad = math.radians(θ)
    return r*math.cos(rad), r*math.sin(rad)


r = 12
θ = 195

x,y = cartesian(r, θ)
print(f"({r}, {θ}°) = ({x:.2f}, {y:.2f})")

```
```bash
$ python3 example.py 
(12, 195°) = (-11.59, -3.11)
$ mypy example.py 
example.py:7: error: Bad number of arguments for type alias, expected: 0, given: 2
Found 1 error in 1 file (checked 1 source file)
```

**What is the actual behavior/output?**
```
example.py:7: error: Bad number of arguments for type alias, expected: 0, given: 2
Found 1 error in 1 file (checked 1 source file)
```

**What is the behavior/output you expect?**
```
Success: no issues found in 1 source file
```
Such as when `from typing import Tuple` is used instead.

**What are the versions of mypy and Python you are using?**

I'm using mypy version 0.77 installed using pip and Python 3.7.7 instead via homebrew on macOS Catalina 10.15.4.

**Do you see the same issue after installing mypy from Git master?**

Yes.

**What are the mypy flags you are using? (For example --strict-optional)**

None. 

**Additional details**

It might be related to #8273. I know `Tuple` and `Callable` are both mentioned in [PEP 484](https://www.python.org/dev/peps/pep-0484/) as being different.

**Comment by hauntsaninja** [comment]
I agree that it's a little weird in the case of Tuple and Callable, but this is the documented behaviour. Because Tuple is a generic type, this is a generic type alias. From https://mypy.readthedocs.io/en/stable/generics.html#generic-type-aliases
> Unsubscripted aliases are treated as original types with free variables replaced with `Any`.
So your `Tuple` is equivalent to `typing.Tuple[typing.Any, ...]`

In your example you seem to be using 2D coordinates, so you could get away with `Tuple = typing.Tuple[S, T]`, which should then make `Tuple[float, float]` work (you might even want to use the same TypeVar, so that your x and y coordinates are of the same type).

We substitute by matching up the type arguments one for one with the free type variables. This feels weird in the case of Tuple, because Tuple can suck up an indeterminate number of type variables. I think if we tried to do something fancier than that, we'd get in trouble with type aliases like `Hmm = Tuple[Tuple, Tuple]`, where we wouldn't know how to distribute the type arguments (what type is `Hmm[int, str, int, str]`). There might still be room for improvement, but I hope this motivates the behaviour you see :-)

**Comment by ufl-taeber** [comment]
Thank you for your response, @hauntsaninja.

Unfortunately, I'm not seeing the behavior you described, which is reaffirming that this is indeed a bug.

You said that aliasing `Tuple` is equivalent to `typing.Tuple[typing.Any, ...]` as per the PEP, but I'm reporting that it is not. Although, I agree that it *should*.

As I reported, I believe `Tuple` is being treated as `typing.Tuple[()]` since the error message is `Bad number of arguments for type alias, expected: 0, given: 2`. If the aliasing were equivalent to `typing.Tuple[typing.Any, ...]` then **there shouldn't be a required number of arguments** as `typing.Any, ...` means any number of arguments can be used. Right?

Also, to clarify, I made up the code in the report to demonstrate the problem. I was having a bit of fun using the non-ASCII `θ` as a variable name; definitely, something I discourage for production use. Anyway, your suggestion of aliasing `Tuple = typing.Tuple[S, T]` doesn't solve the problem generally since another function should be able to use the same `Tuple` alias. Using a different set of arguments, however, would cause an error. For example:
```python
def booleans() -> Tuple[str, bool, bool]:
    return "To be or not to be", False, True
```

Maybe I'm wrong, but, to phrase the report another way:
```python
from typing import Tuple
```
should have the same behavior as
```python
import typing
Tuple = typing.Tuple
```
As I suggested in my report, I workaround the bug by using the `from typing import Tuple` aliasing form. With that form, mypy does equate `Tuple` with `typing.Tuple[typing.Any, ...]` as I'd expect. Still, I think the current interpretation of `Tuple = typing.Tuple` is buggy.

I hope I'm making sense.

Thank you for your time and consideration.

**Comment by hauntsaninja** [comment]
I understand the problem and am sympathetic; it definitely feels unexpected that `from typing import Tuple` and `Tuple = typing.Tuple` behave differently. The last paragraph of my previous comment tries to explain (in my opinion, am not a maintainer), why mypy behaves this way. I won't repeat that here, but let me know if something in that explanation didn't make sense — it's sort of the actual answer to your issue.

Aliasing `Tuple` is in fact equivalent to `typing.Tuple[typing.Any, ...]` and is treated as such by mypy (and this is the documented behaviour, as mentioned above). I think you misunderstand how the latter works. In reference to the part you bolded, tuples of the type `typing.Tuple[typing.Any, ...]` are of indeterminate length (no required length), but the type `typing.Tuple[typing.Any, ...]` cannot itself take any type arguments. This is what the error message is talking about. `typing.Tuple[typing.Any, ...]` is no longer a generic type. Its "contents" are already typed: there's an indeterminate number of them, and each of them is of type `typing.Any`. It's similar to saying `List[Any][int]`; there's an extra type argument and it doesn't mean anything.

Here's some code that helps demonstrate that `Tuple` is equivalent to `typing.Tuple[typing.Any, ...]`:
```
~/delete λ cat test60.py                                                                           130 
import typing

Tuple = typing.Tuple
Tuple2 = typing.Tuple[typing.Any, ...]

def foo() -> Tuple[str, bool, bool]: ...

def bar() -> Tuple2[str, bool, bool]: ...

def baz() -> Tuple: ...

reveal_type(baz())
reveal_type(baz()[100])
~/delete λ mypy test60.py
test60.py:6: error: Bad number of arguments for type alias, expected: 0, given: 3
test60.py:8: error: Bad number of arguments for type alias, expected: 0, given: 3
test60.py:12: note: Revealed type is 'builtins.tuple[Any]'
test60.py:13: note: Revealed type is 'Any'
Found 2 errors in 1 file (checked 1 source file)
```

Multiple functions can in fact use the same `Tuple` alias with `Tuple = typing.Tuple[S, T]`, with different sets of type arguments.
The downside of this approach is that it only works for length-2 tuples; I brought it up because for the example with 2D coordinates, that limitation might actually be nice.

**Comment by ufl-taeber** [comment]
Thank you for expanding on your thinking. It does help me understand your argument that the behavior of mypy is consistent with the documentation and that there is no bug.

I thought I completely understood it, but why would `typing.Dict` act differently when aliased?
```python
import typing

Dict = typing.Dict


def español() -> Dict[bool, str]:
    return {False: "falso", True: "verdadero"}


print(español())
```

If I'm understanding you correctly, the free variables of `typing.Dict` are replaced with `typing.Any`, so using `Dict[bool,str]` should be similar to `typing.Dict[Any, Any][bool,str]`, correct?

When I run this through mypy, however, it reports no errors with this code. Furthermore, if you change `False` to, say `3`, it reports `error: Dict entry 0 has incompatible type "int": "str"; expected "bool": "str"`. Alternatively, changing the return type to `Dict[bool, str, int]` results in `error: "dict" expects 2 type arguments, but 3 given`.

I observe the similar behavior with `List`, `Iterable`, and `Mapping`.

**Comment by hauntsaninja** [comment]
You're right that that is inconsistent with what I interpreted the documentation to be in my earlier comments. Thank you for catching that. mypy treats aliased custom generics as the original type too, so there isn't special casing.

I think my interpretation of the documentation was wrong and just happened to line up with the actual behaviour we observe for Tuple (and Callable); reading it back now, it seems to refer to the use of unsubscripted type aliases, rather than the definition of type aliases without type vars.

As it happens, the code itself does a great job of explaining what's going on, including calling out the exceptional cases of Tuple and Callable.
https://github.com/python/mypy/blob/2c8d76efee7f6419654f0a913d29d9f4edb49574/mypy/nodes.py#L2667

I'm not sure what the resolution is here. It seems hard to support Tuple and Callable, and the type is equivalent to an alias for all other types, which is what you'd want. I'm also not sure what improvements could be made to documentation without confusing things for most users of type aliases. Let me know if you have thoughts.

**Comment by johnpmcconnell** [comment]
@ufl-taeber For your case, I would recommend just importing `Tuple` directly:

```
from typing import Tuple
```

This achieves what you wanted in the first place very directly: avoiding the module prefix in the type annotation.

**Comment by abravalheri** [comment]
Hi guys, just to add to the discussion: I think I run into the same problem when doing:

```python
import sys
if sys.version_info[:2] >= (3, 9):
    Tuple = tuple
else:
    from typing import Tuple
```

This will result in the same error when running mypy on Python 3.9.

**Comment by johnpmcconnell** [comment]
@abravalheri See linked #10068 for a work around

**Comment by abravalheri** [comment]
Thank you very much @johnpmcconnell!
